src/modules/transactions.js

```diff
@@ -0,0 +1,208 @@
+import {ApiRpc} from '../client/api-rpc';
+import uuid from 'uuid';
+import {getRealm} from '../core/realm';
+import {DOCK_TOKEN_UNIT} from '../core/format-utils';
+import {fetchTransactions} from '../core/subscan';
+import BigNumber from 'bignumber.js';
+import {Accounts} from './accounts';
+import { SubstrateNetwork } from './substrate-network';
+
+export const TransactionStatus = {
+  InProgress: 'pending',
+  Failed: 'failed',
+  Complete: 'complete',
+};
+
+/**
+ * Transactions Module
+ */
+export class Transactions {
+  constructor() {
+    this.transactions = [];
+  }
+
+  /**
+   * Get transactions module instance
+   * @returns module instance
+   */
+  static getInstance() {
+    if (!Transactions.instance) {
+      Transactions.instance = new Transactions();
+    }
+
+    return Transactions.instance;
+  }
+  
+  /**
+   * Load external transactions for the given account
+   * 
+   * @param {string} account 
+   */
+  async loadExternalTransactions(account) {
+    const realm = getRealm();
+    const dbTransactions = realm.objects('Transaction').toJSON();
+  
+    const handleTransaction = tx => {
+      if (tx.from !== account && tx.to !== account) {
+        return;
+      }
+  
+      if (dbTransactions.find(item => item.hash === tx.hash)) {
+        return;
+      }
+  
+      const newTx = {
+        amount: BigNumber(tx.amount).times(DOCK_TOKEN_UNIT).toString(),
+        feeAmount: tx.fee,
+        recipientAddress: tx.to,
+        fromAddress: tx.from,
+        id: tx.hash,
+        hash: tx.hash,
+        network: 'mainnet',
+        status: 'complete',
+        date: new Date(parseInt(tx.block_timestamp + '000', 10)),
+      };
+  
+      realm.write(() => {
+        realm.create('Transaction', newTx, 'modified');
+      });
+    };
+  
+    let data;
+    let page = 0;
+  
+    do {
+      data = await fetchTransactions({address: account, page});
+      data.transfers.forEach(handleTransaction);
+      page++;
+    } while (data.hasNextPage);
+  }
+
+  /**
+   * Load transactions for the current accounts
+   * 
+   * @returns transactions
+   */
+  async loadTransactions() {
+    const realm = getRealm();
+    const networkId = SubstrateNetwork.getInstance().networkId;
+
+    if (networkId === 'mainnet') {
+      const accounts = Accounts.getInstance().getAccounts();
+
+      for (const account of accounts) {
+        try {
+          await dispatch(
+            transactionsOperations.loadExternalTransactions(account.id),
+          );
+        } catch (err) {
+          console.error(err);
+        }
+      }
+    }
+
+    let items = realm.objects('Transaction').toJSON();
+
+    if (networkId === 'mainnet') {
+      items = items.filter(item => !(item.status === 'complete' && !item.hash));
+    }
+
+    return items;
+  }
+
+  /**
+   * Update transaction
+   * 
+   * @param {*} transaction 
+   */
+  updateTransaction(transaction){
```

Comment:
can `transaction` be invalid or undefined?

Reply:
realm will handle it, it will throw an error in that case